Assets/GothicVR/Scripts/Npc/Routine.cs

```diff
@@ -7,44 +7,65 @@
 using GVR.Phoenix.Util;
 using GVR.Util;
 using GVR.World;
+using System;
+using System.Data;
 
 namespace GVR.Npc
 {
     public class Routine : MonoBehaviour
     {
-        private static readonly int SPEED = 10;
-        private GameTime gameTime;
+        private const float SPEED = 1f;
+        private GameTime gameTime = new();
+        PxCs.Data.WayNet.PxWayPointData waypoint;
+        RoutineData routine;
+
         public List<RoutineData> routines = new();
+        public Dictionary<string, RoutineData> waypoints = new();
 
-        void Start()
+        private void OnEnable()
         {
-            gameTime = SingletonBehaviour<GameTime>.GetOrCreate();
+            gameTime.minuteChangeCallback.AddListener(lookUpRoutine);
         }
 
-        void Update()
+        private void Start()
         {
-            if (!SingletonBehaviour<DebugSettings>.GetOrCreate().EnableNpcRoutines)
-                return;
-
-            var routine = GetCurrentRoutine();
+            //Initialize first waypoint
+            DateTime time = new(1, 1, 1, 15, 0, 0);
+            lookUpRoutine(time);
+        }
 
+        private void Update()
+        {
+            moveNpc();
+        }
+        private void moveNpc()
+        {
             if (routine == null)
                 return;
-
-            var waypoint = PhoenixBridge.World.waypoints
-                .FirstOrDefault(item => item.name.ToLower() == routine.waypoint.ToLower());
-
+            if (waypoint == null)
+                return;
             var startPosition = gameObject.transform.position;
-            gameObject.transform.position = Vector3.MoveTowards(startPosition, waypoint.position.ToUnityVector(), SPEED * Time.deltaTime);
+            var targetPosition = waypoint.position.ToUnityVector();
+            gameObject.transform.position = Vector3.MoveTowards(startPosition, targetPosition, SPEED * Time.deltaTime);
         }
 
-        private RoutineData GetCurrentRoutine()
+        void lookUpRoutine(DateTime time)
         {
-            var curTime = gameTime.GetCurrentDateTime();
-
-            var routine = routines.FirstOrDefault(item => (item.start <= curTime && curTime < item.stop));
+            if (!SingletonBehaviour<DebugSettings>.GetOrCreate().EnableNpcRoutines)
+                return;
 
-            return routine;
+            getRoutine(time);
+            if (routine == null)
+                return;
+            getWaypoint();
+        }
+        void getRoutine(DateTime time)
```

Comment:
What you do is "setRoutine()" or "calculateRoutine()". But you don't "get" it back.

I think it's fine your calculate the routine time inside and store the current variable. But I would recommend renaming the function.